app/parse.py

```diff
@@ -1,12 +1,160 @@
+
+    return wrapper
+
+
+async def get_page_content(session: ClientSession, url: str) -> str:
+    try:
+        async with session.get(
+                url,
+                timeout=ClientTimeout(total=5)
+        ) as response:
+            response.raise_for_status()
+            return await response.text()
+    except Exception as e:
+        logging.error(f"Error fetching {url}: {e}")
+    return ""
+
+
+def parse_single_course(course: BeautifulSoup) -> Course:
+    modules = course.select("div.CourseModulesList_moduleListItem__HKJqw")
+
+    topics = [
+        (
+            module.select_one(
+                "p.CourseModulesList_topicName__ZrDxT"
+            ).text if module.select_one(
+                "p.CourseModulesList_topicName__ZrDxT") else "N/A",
+            module.select_one(
+                "p.CourseModulesList_topicsCount__yAPxH."
+                "typography_landingTextMain__Rc8BD"
+            ).text if module.select_one(
+                "p.CourseModulesList_topicsCount__yAPxH."
+                "typography_landingTextMain__Rc8BD") else "0",
+        )
+        for module in modules
+    ]
+
+    name = course.select_one(
+        "h1[data-qa='profession-title']"
+    ).text.split(":")[0] if course.select_one(
+        "h1[data-qa='profession-title']"
+    ) else "Unknown"
+    short_description = course.select_one(
+        "p.typography_landingTextMain__Rc8BD"
+    ).text if course.select_one(
+        "p.typography_landingTextMain__Rc8BD"
+    ) else "No description available"
+    duration = course.select_one(
+        ".ComparisonTable_tableBody__W5hzV.mb-24 > "
+        "div:nth-of-type(7) > div:nth-of-type(2)"
+    ).text if course.select_one(
+        ".ComparisonTable_tableBody__W5hzV.mb-24 > "
+        "div:nth-of-type(7) > div:nth-of-type(2)"
+    ) else "Duration not specified"
+
+    return Course(
+        name=name,
+        short_description=short_description,
+        duration=duration,
+        module_count=len(modules),
+        topics=topics,
+    )
+
+
+async def get_courses_urls(session: ClientSession) -> list[str]:
+    page_content = await get_page_content(session, BASE_URL)
+    soup = BeautifulSoup(page_content, "html.parser")
+    return list(
+        set(
+            course["href"].split("/courses/")[-1]
+            for course in soup.select(
+                ".HeaderCoursesDropdown_dropdownWrapper__3Agil "
+                "a[href*='/courses/']"
+            )
+        )
+    )
+
+
+@log_time
+async def get_all_courses() -> list[Any]:
+    async with ClientSession() as session:
+        course_urls = await get_courses_urls(session)
+        tasks = [fetch_course(session, course) for course in course_urls]
+        courses = await asyncio.gather(*tasks)
+
+        valid_courses = [course for course in courses if course is not None]
+
+        return valid_courses
+
+
+async def fetch_course(
+        session: ClientSession,
+        course_slug: str
+) -> Course | None:
```

Comment:
The return type hint `Course | None` is correct, but ensure that `parse_single_course` can handle cases where the HTML structure might not match expectations, potentially returning `None`.